pilot/pkg/proxy/envoy/xds/xds.go

```diff
@@ -139,3 +161,48 @@ func (s *Server) StartGRPC(addr string) error {
 	}()
 	return nil
 }
+
+// ProxyGen implements a proxy generator - any request is forwarded using the agent ADSC connection.
+// Responses are forwarded back on the connection that they are received.
+type ProxyGen struct {
+	adsc   *adsc.ADSC
+	server *Server
+}
+
+// HandleResponse will dispatch a response from a federated
+// XDS server to all connections listening for that type.
+func (p *ProxyGen) HandleResponse(con *adsc.ADSC, res *discovery.DiscoveryResponse) {
+	// TODO: filter the push to only connections that
+	// match a filter.
+	p.server.DiscoveryServer.PushAll(res)
+}
+
+func (s *Server) NewProxy() *ProxyGen {
+	return &ProxyGen{
+		server: s,
+	}
+}
+
+func (p *ProxyGen) AddClient(adsc *adsc.ADSC) {
+	p.server.m.Lock()
+	p.adsc = adsc // TODO: list
+	p.server.m.Unlock()
+}
+
+// TODO: remove clients, multiple clients (agent has only one)
+
+// Generate will forward the request to all remote XDS servers.
+// Responses will be forwarded back to the client.
+//
+// TODO: allow clients to indicate which requests they handle ( similar with topic )
+func (p *ProxyGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	if p.adsc == nil {
+		return nil
+	}
+
+	// TODO: track requests to connections, so resonses from server are dispatched to the right con
+	//
+	_ = p.adsc.Send(w.LastRequest)
```

Comment:
Won't we miss updates if we get two updates before generate is called?

Reply:
Right, missed that case. I added a TODO for now - it requires changing the Generator signature and that's even bigger change. 

For now the proxy is very experimental.